scripts/buildorder.py

```diff
@@ -264,73 +270,49 @@ def read_packages_from_directories(directories, fast_build_mode, full_buildmode)
-                [subpkg.name for subpkg in pkg.subpkgs]
-            )
-
-            if not remaining_deps[other_pkg.name]:  # all deps were already appended?
-                pkg_queue.append(other_pkg)  # should be processed
-
-    if set(pkgs_map.values()) != set(build_order):
-        print("ERROR: Cycle exists. Remaining: ", file=sys.stderr)
-        for name, pkg in pkgs_map.items():
-            if pkg not in build_order:
-                print(name, remaining_deps[name], file=sys.stderr)
-
-        # Print cycles so we have some idea where to start fixing this.
-        def find_cycles(deps, pkg, path):
-            """Yield every dependency path containing a cycle."""
-            if pkg in path:
-                yield path + [pkg]
+    # Start sorting packages by dependencies.
+    while len(pkgs_sort) < len(pkgs_map):
+        for pkg in pkgs_map.values():
+            if pkg.name in pkgs_sort:
+                continue
+            is_subpackage = isinstance(pkg, TermuxSubPackage)
+            if is_subpackage and pkg.parent.name not in pkgs_sort:
+                # Since these packages will be compiled, the sorting should
+                # start with the package and then with its subpackages.
+                pkg = pkg.parent
+                is_subpackage = False
+            subpkgs = [subpkg.name for subpkg in (pkg.parent.subpkgs if is_subpackage else pkg.subpkgs)]
+            for dep in pkg.deps:
+                # First, package dependencies are checked for circular dependencies.
+                # If a package does not have a cyclic dependency, then the condition
+                # is triggered that the dependency is not in the `subpkgs` list and
+                # in the `pkgs_sort` list.  If the condition returns `True`, then the
+                # package is added to the `build_order` list (except for subpackages)
+                # and to the `pkgs_sort` list.
+                for cyclic in cyclic_deps_list:
+                    if pkg.name in cyclic and dep == cyclic[cyclic.index(pkg.name)+1]:
+                        break
+                else:
+                    if dep not in subpkgs and dep not in pkgs_sort:
+                        break
             else:
-                for dep in deps[pkg]:
-                    yield from find_cycles(deps, dep, path + [pkg])
-
-        cycles = set()
-        for pkg in remaining_deps:
-            for path_with_cycle in find_cycles(remaining_deps, pkg, []):
-                # Cut the path down to just the cycle.
-                cycle_start = path_with_cycle.index(path_with_cycle[-1])
-                cycles.add(tuple(path_with_cycle[cycle_start:]))
-        for cycle in sorted(cycles):
-            print(f"cycle: {' -> '.join(cycle)}", file=sys.stderr)
-
-        sys.exit(1)
+                if not is_subpackage:
+                    build_order.append(pkg)
+                pkgs_sort.append(pkg.name)
 
+    # Return sort result
```

Comment:
This is an example of a comment that is antihelpful. `return build_order` couldn't be clearer. It's the overall algorithm that's the problem.

Reply:
ok